File: apps/apple/TeslaSync/feature-views/CostSummaryCards.Model.swift
//
//  CostSummaryCards.Model.swift
//  TeslaSync — P4 feature view · 0111 · CostSummaryCards (Apple)
//
//  State-holder seam (P1/S8) + telemetry seam (P1/S11 diagnostics) + i18n facade
//  (P1/S10). The view binds through `CostSummaryModel`; no networking lives in the view.
//  SwiftUI parity of
//  features/charging/components/cost-analysis/CostSummaryCards.tsx — the charging cost
//  analysis "summary cards" row that renders the aggregated `CoreStats` (total cost, blended
//  $/kWh, cost per distance, total energy, gas savings, savings %) as six `StatBox` tiles.
//  The web component reads its data from the parent `useCostAnalysisData` projection and its
//  display preferences from `useFormatting` (currency symbol) + `useSettings` (gas unit,
//  locale); the production app composes all three into the `CostSummarySource` seam below.
//

import Foundation
import Observation
import OSLog
import SwiftUI

// MARK: - Telemetry seam (P1/S11 diagnostics contract)

/// Emits the `view.opened` product-analytics event for the surface. The default
/// implementation logs via `os.Logger`; the production app injects an adapter that forwards
/// to the shared core `Telemetry.track(.viewOpened(surface:…))`, which is consent-gated and
/// redacted there.
public protocol CostSummaryTelemetry: Sendable {
    func viewOpened(surface: String)
}

/// `os.Logger`-backed default that records the surface open as a `view.opened` event.
public struct OSLogCostSummaryTelemetry: CostSummaryTelemetry {
    private let logger: Logger

    public init() {
        logger = Logger(subsystem: "io.teslasync.app", category: "diagnostics")
    }

    public func viewOpened(surface: String) {
        logger.info("view.opened surface=\(surface, privacy: .public)")
    }
}

// MARK: - State-holder seam (P1/S8 layer)

/// The load lifecycle for the surface's cost-analysis query, mirroring the shared
/// `LoadableState` cases the web source's parent projects from `useCostAnalysisData`
/// (web `isLoading` skeleton / resolved `coreStats` / `coreStats === null` empty / failure).
public enum CostSummaryLoadStatus: Sendable, Equatable {
    case loading
    case loaded
    case empty
    case failed(String)
}

/// Live-stream freshness (ADR-013). Drives the header freshness chip + the cached-data
/// banner so cached tiles are clearly labeled while reconnecting / offline.
public enum CostSummaryConnection: Sendable, Equatable {
    case live
    case stale
    case offline
}

/// The user's gas-unit preference (web `useSettings().settings.gas_unit`). Resolves to the
/// short unit label the web source builds (`gas_unit === 'liter' ? 'L' : 'gal'`); the label
/// is routed through the P1/S10 facade so even the unit abbreviation stays localizable.
public enum CostGasUnit: String, Sendable, Equatable, Codable {
    case gallon
    case liter

    /// The P1/S10 key for the short unit label rendered in the "Gas Savings" subtitle.
    public var labelKey: String {
        self == .liter ? "costAnalysis.stats.gasUnitLiter" : "costAnalysis.stats.gasUnitGallon"
    }

    /// The web English value for the short unit label (`'L'` / `'gal'`).
    public var labelFallback: String {
        self == .liter ? "L" : "gal"
    }
}

/// The aggregated `CoreStats` fields the web source reads. A `nil` snapshot (web
/// `coreStats === null`) renders every tile at the zero sentinel, exactly as the web
/// `coreStats?.field ?? 0` guard does — the grid never renders blank.
public struct CostSummaryStats: Sendable, Equatable {
    /// Web `coreStats.totalCost` — summed `cost_decimal` (display currency).
    public var totalCost: Double
    /// Web `coreStats.count` — number of charging sessions in range (web `fmtInt`).
    public var count: Int
    /// Web `coreStats.avgCostPerKwh` — blended cost per kWh.
    public var avgCostPerKwh: Double
    /// Web `coreStats.costPerDist` — cost per the user's display distance unit.
    public var costPerDist: Double
    /// Web `coreStats.totalEnergy` — total energy added (kWh, already SI-converted upstream).
    public var totalEnergy: Double
    /// Web `coreStats.gallonsEquiv` — gasoline-gallon energy equivalent.
    public var gallonsEquiv: Double
    /// Web `coreStats.savings` — `gasCost - totalCost` (display currency).
    public var savings: Double
    /// Web `coreStats.savingsPercent` — `(savings / gasCost) * 100`.
    public var savingsPercent: Double

    public init(
        totalCost: Double = 0,
        count: Int = 0,
        avgCostPerKwh: Double = 0,
        costPerDist: Double = 0,
        totalEnergy: Double = 0,
        gallonsEquiv: Double = 0,
        savings: Double = 0,
        savingsPercent: Double = 0
    ) {
        self.totalCost = totalCost
        self.count = count
        self.avgCostPerKwh = avgCostPerKwh
        self.costPerDist = costPerDist
        self.totalEnergy = totalEnergy
        self.gallonsEquiv = gallonsEquiv
        self.savings = savings
        self.savingsPercent = savingsPercent
    }

    /// The all-zero snapshot rendered when the cost-analysis query has no rows — the web
    /// `coreStats?.x ?? 0` fallback applied to every field at once.
    public static let zero = CostSummaryStats()
}

/// The display preferences the web source reads outside `coreStats`: the `gasPrice` +
/// `distanceUnit` + `isMiles` props (from the parent page) and the `currencySymbol` /
/// `gas_unit` / `locale` from `useFormatting` + `useSettings`. Bundled so the projection can
/// reproduce every subtitle (`per {unit}`, `vs {price}/{gasUnit}`) and the unit-aware label
/// (`Cost Per {Mile|km}`).
public struct CostSummaryUnitContext: Sendable, Equatable {
    /// Web `gasPrice` prop — the per-unit gasoline price shown in the "Gas Savings" subtitle.
    public var gasPrice: Double
    /// Web `distanceUnit` prop — the display distance symbol shown in the "Cost Per" subtitle.
    public var distanceUnit: String
    /// Web `isMiles` prop — selects the "Mile" vs "km" word in the "Cost Per" label.
    public var isMiles: Bool
    /// Web `useFormatting().currencySymbol` — the currency prefix (defaults to `$`).
    public var currencySymbol: String
    /// Web `useSettings().settings.gas_unit` — resolves the "Gas Savings" unit label.
    public var gasUnit: CostGasUnit
    /// Web global locale (BCP-47) used by `fmtNumber`'s `toLocaleString`; defaults to en-US.
    public var locale: String?

    public init(
        gasPrice: Double = 0,
        distanceUnit: String = "mi",
        isMiles: Bool = true,
        currencySymbol: String = "$",
        gasUnit: CostGasUnit = .gallon,
        locale: String? = nil
    ) {
        self.gasPrice = gasPrice
        self.distanceUnit = distanceUnit
        self.isMiles = isMiles
        self.currencySymbol = currencySymbol
        self.gasUnit = gasUnit
        self.locale = locale
    }
}

/// One coalesced snapshot pushed by a `CostSummarySource`: the cost-analysis load status +
/// the aggregated stats + the display-unit context + the (shared) connection + the in-flight
/// refresh flag + the freshness timestamp.
public struct CostSummaryUpdate: Sendable, Equatable {
    public var status: CostSummaryLoadStatus
    public var stats: CostSummaryStats?
    public var context: CostSummaryUnitContext
    public var refreshing: Bool
    public var connection: CostSummaryConnection
    public var updatedAt: Date?

    public init(
        status: CostSummaryLoadStatus = .loading,
        stats: CostSummaryStats? = nil,
        context: CostSummaryUnitContext = CostSummaryUnitContext(),
        refreshing: Bool = false,
        connection: CostSummaryConnection = .live,
        updatedAt: Date? = nil
    ) {
        self.status = status
        self.stats = stats
        self.context = context
        self.refreshing = refreshing
        self.connection = connection
        self.updatedAt = updatedAt
    }
}

/// The seam the view binds through. The production app implements this over the shared
/// P1/S8 state holders — composing the cost-analysis projection (web `useCostAnalysisData`)
/// with the formatting + settings holders (web `useFormatting` / `useSettings`) and a refresh
/// affordance. Previews + tests use `InMemoryCostSummarySource`. The view never talks to the
/// network directly.
@MainActor
public protocol CostSummarySource: AnyObject {
    var onUpdate: (@MainActor (CostSummaryUpdate) -> Void)? { get set }
    func start()
    func stop()
    /// Re-runs the cost-analysis query from the backend (web `refetch()`).
    func refresh()
}

/// The surface's observable view-model. Subscribes to a `CostSummarySource`, projects the
/// aggregated stats + display context into the six view-ready tiles, and exposes a render
/// `CostSummaryPhase` + freshness for SwiftUI to switch over.
@MainActor
@Observable
public final class CostSummaryModel {
    public private(set) var connection: CostSummaryConnection = .live
    public private(set) var phase: CostSummaryPhase = .loading
    public private(set) var cards: [CostSummaryCardModel] = []
    public private(set) var refreshing = false
    public private(set) var updatedAt: Date?

    @ObservationIgnored private let source: any CostSummarySource
    @ObservationIgnored private let telemetry: any CostSummaryTelemetry
    @ObservationIgnored private var started = false
    @ObservationIgnored private var didAutoRefreshForStale = false

    public init(
        source: any CostSummarySource,
        telemetry: any CostSummaryTelemetry = OSLogCostSummaryTelemetry()
    ) {
        self.source = source
        self.telemetry = telemetry
        source.onUpdate = { [weak self] update in self?.apply(update) }
    }

    /// Begins observing and emits the `view.opened` diagnostics event. Idempotent.
    public func start() {
        guard !started else { return }
        started = true
        telemetry.viewOpened(surface: CostSummaryCards.surfaceSlug)
        source.start()
    }

    /// Stops observing the upstream query.
    public func stop() {
        started = false
        source.stop()
    }

    /// Re-runs the cost-analysis query (web `refetch()`), used by the error-state retry.
    public func refresh() {
        source.refresh()
    }

    private func apply(_ update: CostSummaryUpdate) {
        connection = update.connection
        updatedAt = update.updatedAt
        refreshing = update.refreshing
        cards = CostSummaryProjection.cards(
            from: update.stats,
            context: update.context,
            localize: CostSummaryStrings.string
        )
        phase = CostSummaryProjection.resolvePhase(update.status, hasValue: update.stats != nil)
        handleAutoRefresh(for: update.connection)
    }

    /// Stale → one guarded auto-refresh of the cost-analysis query (prompt "stale chip +
    /// auto-refresh"); reset once live so a later stale episode re-triggers exactly once.
    private func handleAutoRefresh(for connection: CostSummaryConnection) {
        switch connection {
        case .stale:
            guard !didAutoRefreshForStale else { return }
            didAutoRefreshForStale = true
            source.refresh()
        case .live:
            didAutoRefreshForStale = false
        case .offline:
            break
        }
    }
}

/// In-memory source for previews + unit tests. Drive it with `push(_:)`.
@MainActor
public final class InMemoryCostSummarySource: CostSummarySource {
    public var onUpdate: (@MainActor (CostSummaryUpdate) -> Void)?
    public private(set) var startCount = 0
    public private(set) var stopCount = 0
    public private(set) var refreshCount = 0

    private let initial: CostSummaryUpdate?

    public init(initial: CostSummaryUpdate? = nil) {
        self.initial = initial
    }

    public func start() {
        startCount += 1
        if let initial { onUpdate?(initial) }
    }

    public func stop() {
        stopCount += 1
    }

    public func refresh() {
        refreshCount += 1
    }

    /// Pushes a snapshot to the bound model (test/preview affordance).
    public func push(_ update: CostSummaryUpdate) {
        onUpdate?(update)
    }
}

// MARK: - Surface identity + localization facade (P1/S10) — web `t(key, default)`

public extension CostSummaryCards {
    /// Diagnostics surface slug (P1/S11 `view.opened`).
    static let surfaceSlug = "CostSummaryCards"
}

/// Resolves the surface's strings by key with the web English fallback, so the view holds no
/// hardcoded literals. Keys live in the "CostSummaryCards" table, folded into the app
/// `Localizable.xcstrings` catalog at integration time.
public enum CostSummaryStrings {
    public static let table = "CostSummaryCards"

    public static func string(_ key: String, _ fallback: String) -> String {
        NSLocalizedString(key, tableName: table, bundle: .main, value: fallback, comment: "")
    }

    public static func text(_ key: String, _ fallback: String) -> Text {
        Text(verbatim: string(key, fallback))
    }
}

File: apps/android/app/src/main/java/com/teslasync/feature-views/CostSummaryCards/CostSummaryCardsModel.kt
// Pure, framework-free model + projection for the CostSummaryCards feature view — the native analogue of
// everything the web component derives before it returns JSX
// (web/src/features/charging/components/cost-analysis/CostSummaryCards.tsx). No Compose, no Android UI, no
// HTTP: every declaration here is exercised off-device by the :app:testReleaseUnitTest gate, keeping the
// composable a thin render layer.
//
// CostSummaryCards is a presentational surface — the web component takes `coreStats: CoreStats | null`,
// `gasPrice`, `distanceUnit` and `isMiles` as props from the owning CostAnalysisPage (which owns the charging
// query and computes the stats in `useCostAnalysisData`), and reads three context hooks for display:
// `useTranslation` (labels, P1/S10), `useFormatting` (the currency symbol + precision, the native binding of
// the S8 SettingsStore) and `useSettings` (the gas-unit label). It renders six StatBox tiles in a responsive
// 2 / 3 / 6-column grid: Total Cost, Avg $/kWh, Cost Per {Mile|km}, Total Energy, Gas Savings $, Savings %.
//
// Following the sibling SummaryHeroCards port, the owning page threads the computed stats in through the
// shared cache-then-network state-holder layer (P1/S8) as a [UiState]; the [projectUiState] adapter lets the
// composable render every lifecycle state that layer can carry — a loading skeleton grid, a hard error with
// retry, a friendly empty state (the web `!sessions` branch), content, and stale/offline "last known" —
// without ever fetching. The content branch reproduces the web StatBox grid exactly, including the per-card
// glow and icon accent and the composed subtitles.
//
// Number + currency formatting goes through the golden-pinned shared [ChartFormat.number], the native mirror
// of the web `fmtNumber` / `fmtInt` / `fmtWithUnit` (including the web `safeNumber` non-finite to 0 guard).
// The nine `t('costAnalysis.stats.*')` calls resolve from the i18n catalog (P1/S10) at the Compose boundary
// and arrive here as [CostSummaryStrings]. The web source also renders a handful of inline connector/unit
// literals that carry NO i18n key on any platform (the "per " and "vs " connectors, the "gal equiv" / "kWh"
// unit suffixes, the "Mile" / "km" distance words and the "gal" / "L" gas-unit label, the "%" sign and the
// "/" separator); adding catalog keys is outside this surface's allowed-files scope, so — exactly as the
// sibling SummaryHeroCards port composes its symbol-only subtitle from constants — they are reproduced
// verbatim here from documented web-parity constants, never an ad-hoc English string in the view.
//
// `InvalidPackageDeclaration` is suppressed because this surface's mandated directory
// (com/teslasync/feature-views/CostSummaryCards — the P3 prompt's allowed-files path) cannot form a valid
// Kotlin package (a hyphen and a PascalCase segment are illegal in a package identifier), so the package
// intentionally diverges from the path — exactly as the sibling feature-view surfaces do.
// `MatchingDeclarationName` is suppressed for the co-located supporting declarations.
@file:Suppress("MatchingDeclarationName", "InvalidPackageDeclaration")

package io.teslasync.android.featureviews.costsummarycards

import io.teslasync.android.components.charts.ChartFormat
import io.teslasync.android.data.UiPhase
import io.teslasync.android.data.UiState
import io.teslasync.android.data.UnitPreferences
import io.teslasync.shared.core.diagnostics.Logger
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.contentOrNull
import java.util.Locale

// ── Web-parity constants ────────────────────────────────────────────────────────────────────────────

/** Web `formatCurrency(totalCost, 2)` / `formatCurrency(savings, 2)` precision (two fraction digits). */
private const val COST_DECIMALS = 2

/** Web `formatCurrency(avgCostPerKwh, 3)` / `formatCurrency(costPerDist, 3)` precision (three fraction digits). */
private const val RATE_DECIMALS = 3

/** Web `fmtWithUnit(totalEnergy, 'kWh', 1)` / `fmtWithUnit(gallonsEquiv, 'gal equiv', 1)` precision. */
private const val ENERGY_DECIMALS = 1

/** Web `fmtNumber(savingsPercent, 1)` precision (one fraction digit). */
private const val PERCENT_DECIMALS = 1

/** Web `fmtInt(count)` precision (zero fraction digits, locale grouping). */
private const val INT_DECIMALS = 0

/** Energy unit suffix the web source hard-codes (`fmtWithUnit(totalEnergy, 'kWh', 1)`), not converted. */
private const val ENERGY_UNIT = "kWh"

/** Gasoline-equivalent unit suffix the web source hard-codes (`fmtWithUnit(gallonsEquiv, 'gal equiv', 1)`). */
private const val GAL_EQUIV_UNIT = "gal equiv"

/** Trailing percent sign appended to the Savings % value (web `${fmtNumber(savingsPercent, 1)}%`). */
private const val PERCENT_SIGN = "%"

/** Leading connector of the Cost-Per-distance subtitle (web `per ${distanceUnit}`). Trailing space intended. */
private const val PER_PREFIX = "per "

/** Leading connector of the Gas-Savings subtitle (web `vs ${formatCurrency(gasPrice, 2)}/…`). Trailing space. */
private const val VS_PREFIX = "vs "

/** Separator between the gas price and its unit in the Gas-Savings subtitle (web `…/${gasUnitLabel}`). */
private const val PER_UNIT_SEPARATOR = "/"

/** Web `unit: isMiles ? 'Mile' : 'km'` — the imperial distance word substituted into the Cost-Per label. */
private const val MILE_WORD = "Mile"

/** Web `unit: isMiles ? 'Mile' : 'km'` — the metric distance word substituted into the Cost-Per label. */
private const val KM_WORD = "km"

/** Web `settings.gas_unit` document key, read for the gas-unit label. */
private const val KEY_GAS_UNIT = "gas_unit"

/** Web `settings.gas_unit === 'liter'` sentinel value. */
private const val GAS_UNIT_LITER = "liter"

/** Web `settings.gas_unit === 'liter' ? 'L' : 'gal'` — the litre gas-unit label. */
private const val GAS_LABEL_LITER = "L"

/** Web `settings.gas_unit === 'liter' ? 'L' : 'gal'` — the (default) gallon gas-unit label. */
private const val GAS_LABEL_GALLON = "gal"

private const val DEFAULT_CURRENCY = "$"
private const val DEFAULT_PRECISION = 2
private const val DEFAULT_LOCALE_TAG = "en-US"
private const val KEY_CURRENCY_SYMBOL = "currency_symbol"

// ── Inputs ──────────────────────────────────────────────────────────────────────────────────────────

/**
 * The eight `CoreStats` fields the web CostSummaryCards reads off its `coreStats` prop — the subset of the
 * larger web `CoreStats` type (web/src/features/charging/components/cost-analysis/types.ts) that THIS surface
 * renders. All monetary/ratio figures are doubles mirroring the web `number` shape; [count] is the integer
 * session count (web `coreStats.count`, rendered with `fmtInt`). The owning page computes them as finite sums,
 * and the formatters additionally apply the web `safeNumber` non-finite to 0 guard.
 *
 * @property totalCost summed session cost (web `coreStats.totalCost`).
 * @property count number of charging sessions (web `coreStats.count`).
 * @property avgCostPerKwh blended cost per kWh (web `coreStats.avgCostPerKwh`).
 * @property costPerDist cost per user-preferred distance unit (web `coreStats.costPerDist`).
 * @property totalEnergy total energy added, kWh (web `coreStats.totalEnergy`).
 * @property gallonsEquiv gasoline-gallon energy equivalent (web `coreStats.gallonsEquiv`).
 * @property savings money saved versus gasoline (web `coreStats.savings`).
 * @property savingsPercent savings as a percentage of the gas cost (web `coreStats.savingsPercent`).
 */
data class CostSummaryStats(
    val totalCost: Double,
    val count: Int,
    val avgCostPerKwh: Double,
    val costPerDist: Double,
    val totalEnergy: Double,
    val gallonsEquiv: Double,
    val savings: Double,
    val savingsPercent: Double,
)

/**
 * The full prop bundle the owning page threads into this surface — the web component's `coreStats` plus the
 * `gasPrice`, `distanceUnit` and `isMiles` props, grouped so the host has a single value to carry through the
 * [UiState].
 *
 * @property stats the eight figures the six cards render (web `coreStats`, non-null in the content state).
 * @property gasPrice the gas price the Savings subtitle compares against (web `gasPrice`).
 * @property distanceUnit the user's distance-unit abbreviation, `mi` or `km` (web `distanceUnit`).
 * @property isMiles whether the user prefers miles, selecting the Cost-Per label's distance word (web `isMiles`).
 */
data class CostSummarySnapshot(
    val stats: CostSummaryStats,
    val gasPrice: Double,
    val distanceUnit: String,
    val isMiles: Boolean,
)

// ── Render-ready projection types ──────────────────────────────────────────────────────────────────

/** The web StatBox `glow` prop — tints the panel border accent (web GlassPanel glow). */
enum class CostStatGlow { Cyan, Green, None }

/** The web icon color (`text-{color}-400`) applied as the tile glyph tint. */
enum class CostStatIconTone { Cyan, Yellow, Blue, Green, Red, Emerald }

/** Which authored lucide glyph a tile carries (web `icon`), resolved to an ImageVector in the composable. */
enum class CostStatIcon { Dollar, Zap, Car, Fuel, TrendingDown }

/**
 * One fully resolved StatBox tile — the native analogue of a single web `<StatBox>` invocation. Pure data (no
 * Compose types) so the whole projection is asserted off-device. The [label] is already localized (resolved
 * from the i18n catalog at the Compose boundary and handed in via [CostSummaryStrings]); the [sub] is the
 * composed secondary line.
 *
 * @property label the localized tile label (web `label`).
 * @property value the formatted primary value, unit/symbol included (web `value`).
 * @property sub the secondary line, or `null` when the web omits it (web `sub`).
 * @property glow the panel border accent (web `glow`).
 * @property iconTone the glyph tint (web icon `text-{color}-400`).
 * @property icon the glyph slot (web `icon`).
 */
data class CostSummaryCard(
    val label: String,
    val value: String,
    val sub: String?,
    val glow: CostStatGlow,
    val iconTone: CostStatIconTone,
    val icon: CostStatIcon,
)

/**
 * The localized strings the composable resolves once (P1/S10) and threads into the projection so the
 * render-ready cards carry no English literal. Keys map 1:1 to the web `t('costAnalysis.stats.*')` calls;
 * [costPerDistTemplate] is the raw "Cost Per %1$s" resource into which the projection substitutes the
 * distance word.
 */
data class CostSummaryStrings(
    val totalCost: String,
    val sessions: String,
    val avgPerKwh: String,
    val blendedRate: String,
    val costPerDistTemplate: String,
    val totalEnergy: String,
    val gasSavings: String,
    val savingsPercent: String,
    val vsGasoline: String,
)

/**
 * The display preferences this surface resolves from the live `/settings` document — the native union of the
 * web `useFormatting` reads (currency symbol + precision), the `fmtNumber` global locale, and the
 * `useSettings` gas-unit label. Resolved from one settings document, mirroring the web hooks which all derive
 * from `useSettings`.
 *
 * @property currencySymbol the resolved currency symbol with the web blank/whitespace to "$" fallback applied.
 * @property precision the default fraction digits (web `useFormatting` `userPrecision`); the cards override it
 *   with explicit decimals per the web calls, so this is the cold-start default only.
 * @property locale the BCP-47 locale driving number grouping/separators (web `fmtNumber` global locale).
 * @property gasUnitLabel the gas-unit label, `gal` or `L` (web `settings.gas_unit === 'liter' ? 'L' : 'gal'`).
 */
data class CostSummaryDisplayPrefs(
    val currencySymbol: String,
    val precision: Int,
    val locale: Locale,
    val gasUnitLabel: String,
) {
    companion object {
        /** The "$", 2-dp, en-US, "gal" defaults applied before settings load (web cold-start defaults). */
        val DEFAULT: CostSummaryDisplayPrefs = from(null)

        /** Resolves the currency + precision + locale + gas-unit preferences from one `/settings` document. */
        fun from(settings: JsonElement?): CostSummaryDisplayPrefs {
            val unitPref = UnitPreferences.fromSettings(settings)
            val obj = settings as? JsonObject
            val rawSymbol = obj.stringOrNull(KEY_CURRENCY_SYMBOL)
            return CostSummaryDisplayPrefs(
                currencySymbol = if (!rawSymbol.isNullOrBlank()) rawSymbol else DEFAULT_CURRENCY,
                precision = unitPref.precision ?: DEFAULT_PRECISION,
                locale = localeFor(unitPref.locale),
                gasUnitLabel = if (obj.stringOrNull(KEY_GAS_UNIT) == GAS_UNIT_LITER) GAS_LABEL_LITER else GAS_LABEL_GALLON,
            )
        }

        private fun localeFor(tag: String?): Locale = Locale.forLanguageTag(tag?.takeIf { it.isNotBlank() } ?: DEFAULT_LOCALE_TAG)
    }
}

// ── Projection ──────────────────────────────────────────────────────────────────────────────────────

/**
 * Pure projection from the surface's prop + display preferences to its render-ready cards — a 1:1 port of the
 * derivations the web component performs. The composable resolves [CostSummaryStrings] and
 * [CostSummaryDisplayPrefs] from the i18n catalog and the live settings, then hands them here.
 */
object CostSummaryCardsProjection {
    /**
     * Maps the host's `(snapshot, isLoading)` onto the shared cache-then-network [UiState] (P1/S8): loading
     * wins outright (skeleton chrome), a present snapshot renders [UiPhase.Content], and an absent snapshot
     * renders [UiPhase.Empty] — the native expression of the web `if (!sessions || sessions.length === 0)`
     * empty branch (no charging data, so no `coreStats`). The host's stateful binding can additionally carry
     * refreshing/stale/offline/error, which the composable renders too.
     */
    fun projectUiState(
        snapshot: CostSummarySnapshot?,
        isLoading: Boolean,
    ): UiState<CostSummarySnapshot> =
        when {
            isLoading -> UiState.loading()
            snapshot != null -> UiState(phase = UiPhase.Content, data = snapshot)
            else -> UiState(phase = UiPhase.Empty)
        }

    /**
     * The six StatBox tiles in web source order. Each value/subtitle reproduces the matching web call — the
     * `formatCurrency` costs, the `fmtWithUnit` energy, the `fmtNumber` percentage, the composed
     * `${fmtInt(count)} sessions` / `per ${distanceUnit}` / `vs ${formatCurrency(gasPrice, 2)}/${gasUnitLabel}`
     * subtitles — formatted for [prefs] (locale grouping + the user's currency + gas-unit label), with the
     * verbatim glow and icon accent the web hard-codes per tile.
     */
    fun cards(
        snapshot: CostSummarySnapshot,
        prefs: CostSummaryDisplayPrefs,
        strings: CostSummaryStrings,
    ): List<CostSummaryCard> {
        val s = snapshot.stats
        val locale = prefs.locale
        return listOf(
            CostSummaryCard(
                label = strings.totalCost,
                value = formatCurrency(s.totalCost, prefs, COST_DECIMALS),
                sub = fmt(s.count.toDouble(), INT_DECIMALS, locale) + " " + strings.sessions, // parity:allow Int to Double widening
                glow = CostStatGlow.Cyan,
                iconTone = CostStatIconTone.Cyan,
                icon = CostStatIcon.Dollar,
            ),
            CostSummaryCard(
                label = strings.avgPerKwh,
                value = formatCurrency(s.avgCostPerKwh, prefs, RATE_DECIMALS),
                sub = strings.blendedRate,
                glow = CostStatGlow.None,
                iconTone = CostStatIconTone.Yellow,
                icon = CostStatIcon.Zap,
            ),
            CostSummaryCard(
                label = costPerDistanceLabel(strings.costPerDistTemplate, snapshot.isMiles, locale),
                value = formatCurrency(s.costPerDist, prefs, RATE_DECIMALS),
                sub = PER_PREFIX + snapshot.distanceUnit,
                glow = CostStatGlow.None,
                iconTone = CostStatIconTone.Blue,
                icon = CostStatIcon.Car,
            ),
            CostSummaryCard(
                label = strings.totalEnergy,
                value = withUnit(s.totalEnergy, ENERGY_UNIT, ENERGY_DECIMALS, locale),
                sub = withUnit(s.gallonsEquiv, GAL_EQUIV_UNIT, ENERGY_DECIMALS, locale),
                glow = CostStatGlow.Green,
                iconTone = CostStatIconTone.Green,
                icon = CostStatIcon.Zap,
            ),
            CostSummaryCard(
                label = strings.gasSavings,
                value = formatCurrency(s.savings, prefs, COST_DECIMALS),
                sub = gasSavingsSub(snapshot.gasPrice, prefs),
                glow = CostStatGlow.Green,
                iconTone = CostStatIconTone.Red,
                icon = CostStatIcon.Fuel,
            ),
            CostSummaryCard(
                label = strings.savingsPercent,
                value = fmt(s.savingsPercent, PERCENT_DECIMALS, locale) + PERCENT_SIGN,
                sub = strings.vsGasoline,
                glow = CostStatGlow.Green,
                iconTone = CostStatIconTone.Emerald,
                icon = CostStatIcon.TrendingDown,
            ),
        )
    }

    /**
     * The Gas-Savings subtitle — the web `vs ${formatCurrency(gasPrice, 2)}/${gasUnitLabel}`: the "vs "
     * connector, the gas price at two decimals, the "/" separator and the resolved gas-unit label.
     */
    fun gasSavingsSub(
        gasPrice: Double,
        prefs: CostSummaryDisplayPrefs,
    ): String = VS_PREFIX + formatCurrency(gasPrice, prefs, COST_DECIMALS) + PER_UNIT_SEPARATOR + prefs.gasUnitLabel

    /**
     * The Cost-Per tile label — the web `t('costAnalysis.stats.costPerDist', { unit: isMiles ? 'Mile' : 'km' })`.
     * Substitutes the imperial/metric distance word into the raw "Cost Per %1$s" [template] using [locale] so a
     * locale-specific format string still resolves correctly.
     */
    fun costPerDistanceLabel(
        template: String,
        isMiles: Boolean,
        locale: Locale = Locale.getDefault(),
    ): String = String.format(locale, template, if (isMiles) MILE_WORD else KM_WORD)

    /**
     * Formats a currency [amount] the way the web `useFormatting().formatCurrency` does — the resolved
     * [CostSummaryDisplayPrefs.currencySymbol] followed by a grouped number via the shared [ChartFormat.number].
     * [decimals] defaults to the user's precision; the cards pass explicit decimals per the web calls.
     */
    fun formatCurrency(
        amount: Double,
        prefs: CostSummaryDisplayPrefs,
        decimals: Int = prefs.precision,
        locale: Locale = prefs.locale,
    ): String = prefs.currencySymbol + fmt(amount, decimals, locale)

    /**
     * A grouped number at [decimals] fraction digits — the web `fmtNumber`, including its `safeNumber` guard
     * (a non-finite value renders as 0 rather than the [ChartFormat] em-dash, matching the web output).
     */
    private fun fmt(
        value: Double,
        decimals: Int,
        locale: Locale,
    ): String = ChartFormat.number(if (value.isFinite()) value else 0.0, decimals.coerceAtLeast(0), locale)

    /** A grouped number with a trailing [unit] (web `fmtWithUnit`): `${fmtNumber(value, decimals)} ${unit}`. */
    private fun withUnit(
        value: Double,
        unit: String,
        decimals: Int,
        locale: Locale,
    ): String = fmt(value, decimals, locale) + " " + unit
}

// ── Diagnostics (P1/S11) ─────────────────────────────────────────────────────────────────────────────

/**
 * The one PII-safe diagnostic this surface emits (P1/S11). Carries only the surface [SLUG] — never a cost,
 * energy, savings, session count, or gas-price figure — so a diagnostics line can never leak fleet usage or
 * charging economics.
 */
object CostSummaryCardsDiagnostics {
    /** Diagnostics surface slug emitted with the `view.opened` event. */
    const val SLUG: String = "CostSummaryCards"

    private const val VIEW_OPENED = "view.opened"
    private const val SURFACE_KEY = "surface"

    /** Emits the `view.opened` diagnostic for this surface. Call from the composable's first-composition effect. */
    fun recordViewOpened(logger: Logger) {
        logger.info(VIEW_OPENED, mapOf(SURFACE_KEY to SLUG))
    }
}

// ── JSON decode helper (web blank/whitespace → "$" parity) ────────────────────────────────────────────

private fun JsonObject?.stringOrNull(key: String): String? = (this?.get(key) as? JsonPrimitive)?.contentOrNull
